pkg/common_controller/snapshot_controller.go

```diff
@@ -258,51 +300,217 @@ func (ctrl *csiSnapshotController) syncUnreadySnapshot(snapshot *crdv1.VolumeSna
+		newContent, err := ctrl.checkandBindSnapshotContent(snapshot, content)
 		if err != nil {
 			// snapshot is bound but content is not bound to snapshot correctly
 			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotBindFailed", fmt.Sprintf("Snapshot failed to bind VolumeSnapshotContent, %v", err))
 			return fmt.Errorf("snapshot %s is bound, but VolumeSnapshotContent %s is not bound to the VolumeSnapshot correctly, %v", uniqueSnapshotName, content.Name, err)
 		}
-		// snapshot is already bound correctly, check the status and update if it is ready.
-		klog.V(5).Infof("Check and update snapshot %s status", uniqueSnapshotName)
-		if err = ctrl.checkandUpdateBoundSnapshotStatus(snapshot, contentBound); err != nil {
-			return err
+
+		// update snapshot status
+		if newContent.Status != nil {
+			// Update snapshot status with creationTime
+			for i := 0; i < ctrl.createSnapshotContentRetryCount; i++ {
+				klog.V(5).Infof("syncUnreadySnapshot [%s]: trying to update snapshot status", utils.SnapshotKey(snapshot))
+				_, err = ctrl.updateSnapshotStatus(snapshot, newContent)
+				if err == nil {
+					break
+				}
+				klog.V(4).Infof("failed to update snapshot %s status: %v", utils.SnapshotKey(snapshot), err)
+			}
+
+			if err != nil {
+				// update snapshot status failed
+				ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotStatusUpdateFailed", fmt.Sprintf("Snapshot status update failed, %v", err))
+				return err
+			}
 		}
+
 		return nil
-	} else { // snapshot.Source.Spec.VolumeSnapshotContentName == nil
-		// find a matching volume snapshot content
+	} else { // snapshot.Spec.Source.VolumeSnapshotContentName == nil - dynamically creating snapshot
+		klog.V(5).Infof("before getMatchSnapshotContent for snapshot %s", uniqueSnapshotName)
 		if contentObj := ctrl.getMatchSnapshotContent(snapshot); contentObj != nil {
 			klog.V(5).Infof("Find VolumeSnapshotContent object %s for snapshot %s", contentObj.Name, uniqueSnapshotName)
-			if err := ctrl.checkandUpdateBoundSnapshotStatus(snapshot, contentObj); err != nil {
+			newSnapshot, err := ctrl.bindandUpdateVolumeSnapshot(contentObj, snapshot)
+			if err != nil {
 				return err
 			}
-			klog.V(5).Infof("checkandUpdateBoundSnapshotStatus %v", snapshot)
+			klog.V(5).Infof("bindandUpdateVolumeSnapshot %v", newSnapshot)
+			return nil
 		} else if snapshot.Status != nil && snapshot.Status.BoundVolumeSnapshotContentName != nil {
 			contentObj, found, err := ctrl.contentStore.GetByKey(*snapshot.Status.BoundVolumeSnapshotContentName)
 			if err != nil {
 				return err
 			}
 			if !found {
-				ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotContentMissing", "VolumeSnapshotContent is missing")
-				return fmt.Errorf("snapshot %s is bound to a non-existing content %s", uniqueSnapshotName, *snapshot.Status.BoundVolumeSnapshotContentName)
+				// TODO(xyang): this is not an error now because
```

Comment:
We can't make this an error now because it is a normal condition when content is deleted before the snapshot.  We will see this code path while deleting a snapshot.

Reply:
we can make it an error if the snapshot.DeletionTimestamp is unset? the deletion workflow will be triggered only when a snapshot has been deleted.